src/main/scala/viper/gobra/frontend/info/implementation/resolution/Enclosing.scala

```diff
@@ -30,6 +30,39 @@ trait Enclosing { this: TypeInfoImpl =>
     case _ => id
   })
 
+  lazy val enclosingLoop: PNode => Option[PForStmt] = {
+    down[Option[PForStmt]](None){ case x: PForStmt => Some(x) }
+  }
+
+  // Returns the enclosing loop that has a specific label
+  // It also returns the invariants of that loop
+  def enclosingLabeledLoop(label: PLabelUse, node: PNode) : Option[PForStmt] = {
+    enclosingLoop(node) match {
+      case None => None
+      case Some(encLoop) => encLoop match {
+        case tree.parent(l: PLabeledStmt) if l.label.name == label.name => Some(encLoop)
+        case tree.parent(p) => enclosingLabeledLoop(label, p)
+        case _ => violation("No parent found for a loop statement.")
+      }
+    }
+  }
+
+  // Returns the number of loops between a node 'node' and a label 'label'
+  // Used when 'continue l' statements that correspond to
+  // a for statement with label l are being desugared. This number will be
+  // used to fetch the unique label identifier for that for loop where
+  // the continue statement must goto.
+  def enclosingLoopOrder(label: PLabelUse, node: PNode, order: Int = 0) : Int = {
```

Comment:
Is the purpose of `order` here to be able to increment its value in the recursive call to itself? Is that the only purpose of this param? If so, I usually prefer to introduce an auxiliary method that takes that extra param inside the definifition of the method to avoid cluttering the signature of the method. Signature of public methods should only have parameters with a clear meaning outside of the implementation. Here, this is less of a problem given that the signature that you expose in the interface `ExternalTypeInfo` don't have this parameter.